storage/test/sparse_merkle/internal_node_test.cpp

```diff
@@ -0,0 +1,373 @@
+  std::copy(key1_hash.data(), key1_hash.data() + key1_hash.size(), key2_hash_data.begin());
+  key2_hash_data[0] ^= 0x40;
+  auto key2_hash = Hash(key2_hash_data);
+
+  ASSERT_EQ(1, key1_hash.bits_in_common(key2_hash));
+
+  auto value1_hash = hasher.hash(value1, strlen(value1));
+  auto value2_hash = hasher.hash(value2, strlen(value2));
+  auto leaf_key1 = LeafKey(key1_hash, Version(1));
+  auto leaf_key2 = LeafKey(key2_hash, Version(2));
+  auto child1 = LeafChild{value1_hash, leaf_key1};
+  auto child2 = LeafChild{value2_hash, leaf_key2};
+
+  // The key of the first child in the 4-level BatchedInternalNode tree.
+  Nibble child1_key = key1_hash.getNibble(depth);
+
+  auto result = node.insert(child1, depth);
+  result = node.insert(child2, depth);
+
+  // No nodes were overwritten and no new BatchedInternalNodes created
+  ASSERT_TRUE(std::holds_alternative<BatchedInternalNode::InsertComplete>(result));
+  auto successful_insert = std::get<BatchedInternalNode::InsertComplete>(result);
+  ASSERT_FALSE(successful_insert.stale_leaf.has_value());
+
+  ASSERT_EQ(4, node.numChildren());
+  // The root + one internal node below the root
+  ASSERT_EQ(2, node.numInternalChildren());
+  ASSERT_EQ(2, node.numLeafChildren());
+
+  // The version should be 2.
+  ASSERT_EQ(Version(2), node.version());
+
+  Hash level_3_parent_hash;
+  if (child1_key.getBit(2)) {
+    // The key was inserted at the right child of the InternalChild at height 3
+    level_3_parent_hash = hasher.parent(value2_hash, value1_hash);
+  } else {
+    // The key was inserted at the left child of the InternalChild at height 3
+    level_3_parent_hash = hasher.parent(value1_hash, value2_hash);
+  }
+
+  Hash root_hash;
+  if (child1_key.getBit(3)) {
+    // The internal node is at the right of the root
+    root_hash = hasher.parent(PLACEHOLDER_HASH, level_3_parent_hash);
+  } else {
+    root_hash = hasher.parent(level_3_parent_hash, PLACEHOLDER_HASH);
+  }
+
+  ASSERT_EQ(node.hash(), root_hash);
+}
+
+/*
+// The logical tree inside the BatchedInternalNode looks like the following at the end of this test
+//
+//                               Root
+//                                 |
+//                         -----------------
+//                         |               |
+//                      Internal      Placeholder
+//                         |
+//                 ----------------
+//                 |              |
+//              Internal       Placeholder
+//                 |
+//          ----------------
+//          |              |
+//      Internal       Placeholder
+//          |
+//    ----------------
+//    |              |
+// Internal       Placeholder
+//
+*/
+
+TEST(internal_node_tests, split_until_new_batch_node_needed) {
+  BatchedInternalNode node;
+
+  Hasher hasher;
+  const char* key1 = "artist";
+  const char* value1 = "REM";
+  const char* value2 = "Nas";
+  size_t depth = 0;
+
+  auto key1_hash = hasher.hash(key1, strlen(key1));
```

Comment:
I've been debugging and, if not mistaken, the comment about the structure of the tree having all internal children being left ones seems incorrect, because some of the bits in the first nibble of `key1_hash` are 1 and that leads to `insertInternalChildren()` adding right children. Maybe we can change `key1_hash` to reflect the comment as:

```
  auto key1_hash = hasher.hash(key1, strlen(key1));
  // Copy the actual hash and set the first nibble to zeroes to make all
  // internal children left ones.
  std::array<uint8_t, Hash::SIZE_IN_BYTES> key1_hash_data;
  std::copy(key1_hash.data(), key1_hash.data() + key1_hash.size(), key1_hash_data.begin());
  key1_hash_data[0] &= 0x0F;
  key1_hash = Hash{key1_hash_data};
```

Reply:
Great catch! That's a good point. I wasn't really thinking about that. I was more worried that we hit height 0 in this test. I Will change as you suggest, since this makes the diagram work and doesn't change the intent of the test.